### PR TITLE
Flag added to show tree of all namespaces

### LINKED ISSUE #130: HNC: allow `kubectl hnc tree` to show _all_ namespaces if no argument is given
(Filed as a result of #93)

Currently `kubectl hnc tree` requires an argument, which is the root namespace of the tree to display. We should also allow the user to omit the root argument and print _every_ namespace (and their subnamespaces) instead.

/good-first-issue

### PR DESCRIPTION
**Pull request by shivi28**
What this PR does / why we need it:
This PR adds a feature that allow kubectl hnc tree --all-namespaces to show all namespaces. This feature enables users to list all namespaces along with their heirarchies.

Suggested changes :
`kubectl hnc tree foo # displays foo`
`kubectl hnc tree # error`
`kubectl hnc tree --all-namespaces # displays the tree of all namespaces`
`kubectl hnc tree -A # synonym for --all-namespaces`

Which issue(s) this PR fixes:
Fixes #130

### CODE REVIEW

File: incubator/hnc/pkg/kubectl/tree.go [hunk 1]
```diff
@@ -17,13 +17,14 @@ package kubectl
...
+	"os"
 	"sort"
 	"strconv"
 	"strings"
 
-	"github.com/spf13/cobra"
-
 	api "github.com/kubernetes-sigs/multi-tenancy/incubator/hnc/api/v1alpha1"
+	"github.com/spf13/cobra"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
```
Comment (adrianludwin) [inline incubator/hnc/pkg/kubectl/tree.go]: Usually we have up to three "groups" of imports:

- Standard library ("os", "strings", etc)
- Packages from other projects ("cobra", "metav1", etc)
- Packages from this project ("api")

I won't hold this PR up for this change, but I'll probably change it back myself.

File: incubator/hnc/pkg/kubectl/tree.go [hunk 2]
```diff
@@ -32,19 +33,35 @@ var (
 )
 
 var treeCmd = &cobra.Command{
-	Use:   "tree",
-	Short: "Displays the hierarchy tree rooted at the given namespace",
-	Args:  cobra.ExactArgs(1),
+	Use:     "tree",
+	Short:   "Displays the hierarchy tree rooted at the given namespaces separated by space and if no namespace is provided then throws an error, and in case of --all-namespace/-A flag shows tree for all root namespaces",
```
Comment (adrianludwin) [inline incubator/hnc/pkg/kubectl/tree.go]: This isn't really a short description any more - I'd change this to "Display one or more hierarchy trees." If you add help text to `-A`, then saying `kubectl hnc help tree` _should_ give enough information to make it clear.

File: incubator/hnc/pkg/kubectl/tree.go [hunk 3]
```diff
@@ -32,19 +33,35 @@ var (
...
+	Aliases: []string{"tree"},
```
Comment (adrianludwin) [inline incubator/hnc/pkg/kubectl/tree.go]: I don't think you need this? The command is already called "tree," why did you add this line?

File: incubator/hnc/pkg/kubectl/tree.go [hunk 4]
```diff
@@ -93,5 +110,20 @@ func nameAndFootnotes(hier *api.HierarchyConfiguration) string {
 }
 
 func newTreeCmd() *cobra.Command {
+	treeCmd.Flags().CountP("all-namespaces", "A", "Show tree for all namespaces")
```
Comment (adrianludwin) [inline incubator/hnc/pkg/kubectl/tree.go]: Maybe "Displays all trees on the cluster"?

File: incubator/hnc/pkg/kubectl/tree.go [hunk 5]
```diff
@@ -32,19 +33,35 @@ var (
...
-		printSubtree("", hier)
-
+		var nsList []string
+		defaultList := len(args) == 0
+		nsList = args
+		if flags.Changed("all-namespaces") {
+			nsList = getAllNamespaces()
+		}
+		if defaultList && !flags.Changed("all-namespaces") {
+			fmt.Printf("Error: requires some argument or flag but provided none\n")
```
Comment (adrianludwin) [inline incubator/hnc/pkg/kubectl/tree.go]: You should give more guidance than this. Maybe "you must specify the root of the tree(s) to display or else specify --all-namespaces"?

File: incubator/hnc/pkg/kubectl/tree.go [hunk 6]
```diff
@@ -93,5 +110,20 @@ func nameAndFootnotes(hier *api.HierarchyConfiguration) string {
 }
 
 func newTreeCmd() *cobra.Command {
+	treeCmd.Flags().BoolP("all-namespaces", "A", true, "Displays all trees on the cluster")
```
Comment (rjbez17) [inline incubator/hnc/pkg/kubectl/tree.go]: Doesn't this default the flag to true?

Comment (shivi28) [reply incubator/hnc/pkg/kubectl/tree.go]: Yes default value is true, Should I make it false @rjbez17 ? Actually I am not clear with its context :(

Comment (adrianludwin) [reply incubator/hnc/pkg/kubectl/tree.go]: Good catch Ryan. Yes, the default should be false. If the default is true, then `kubectl hnc tree` will be the same as `kubectl hnc tree -A` which is _not_ what we want.

File: incubator/hnc/pkg/kubectl/tree.go [hunk 7]
```diff
@@ -33,18 +35,33 @@ var (
...
 	Run: func(cmd *cobra.Command, args []string) {
-		nnm := args[0]
+		flags := cmd.Flags()
 		footnotesByMsg = map[string]int{}
 		footnotes = []string{}
-		hier := getHierarchy(nnm)
-		fmt.Println(nameAndFootnotes(hier))
-		printSubtree("", hier)
-
+		var nsList []string
```
Comment (adrianludwin) [inline incubator/hnc/pkg/kubectl/tree.go]: You don't need this line; on line 45 you could just say:

`nsList := args`

which initializes the new variable to the same type as `args`.

File: incubator/hnc/pkg/kubectl/tree.go [hunk 8]
```diff
@@ -33,18 +35,33 @@ var (
...
+		defaultList := len(args) == 0
+		nsList = args
+		if flags.Changed("all-namespaces") {
+			nsList = getAllNamespaces()
+		}
+		if defaultList && !flags.Changed("all-namespaces") {
+			fmt.Printf("Error: Must specify the root of the tree(s) to display or else specify --all-namespaces\n")
+			return
```
Comment (adrianludwin) [inline incubator/hnc/pkg/kubectl/tree.go]: This should probably be `os.Exit(1)` to indicate to any calling script that there was an error.

**Comment by adrianludwin** [comment]
I think we resolved all of Ryan's concerns so I'm going to approve, we can always have another go at this if there are more problems.

/approve
